File: app/src/main/java/com/tuanlvt/mvp_architecture/data/source/MovieRepository.java
package com.tuanlvt.mvp_architecture.data.source;

import com.tuanlvt.mvp_architecture.data.model.BaseResponse;
import com.tuanlvt.mvp_architecture.data.model.Movie;
import io.reactivex.Observable;
import java.util.List;

/**
 * Created by FRAMGIA\le.vu.tan.tuan on 24/08/2018.
 * tantuan127@gmail.com
 */
public class MovieRepository {

    private MovieDataSource.RemoteDataSource mRemoteDataSource;
    private MovieDataSource.LocalDataSource mLocalDataSource;

    public MovieRepository(MovieDataSource.RemoteDataSource remoteDataSource,
            MovieDataSource.LocalDataSource localDataSource) {
        mRemoteDataSource = remoteDataSource;
        mLocalDataSource = localDataSource;
    }

    public Observable<BaseResponse<List<Movie>>> getData() {
        return mRemoteDataSource.getData();
    }
}

File: app/src/main/java/com/tuanlvt/mvp_architecture/data/source/MovieRepository.kt
package com.tuanlvt.mvp_architecture.data.source

import com.tuanlvt.mvp_architecture.data.model.Movie
import com.tuanlvt.mvp_architecture.data.source.remote.OnDataResultListener

class MovieRepository private constructor(private val remote: MovieDataSource.Remote,
                                          private val local: MovieDataSource.Local) {
    fun getMovie(listener: OnDataResultListener<MutableList<Movie>>) {
        remote.getMovies(listener)
    }

    companion object {
        private var instance: MovieRepository? = null

        fun getInstance(remote: MovieDataSource.Remote,
                        local: MovieDataSource.Local) = synchronized(this) {
            instance ?: MovieRepository(remote, local).also { instance = it }
        }
    }
}
